feat(electric): JWT expiration

kevin-dp opened the pull request:
When the client authenticates, Electric sets a timer that expires on the time specified by the token's "exp" (expiration) claim. When that timer is fired, the client's socket is closed with a status code 4000 and reason `"JWT-expired"`. This fixes VAX-691, VAX-1434 and VAX-1478.

The TypeScript client is also modified to handle socket close events due to JWT expiration. When the socket is closed because the JWT expired, this is signaled to the developer through an AuthStateChange through the notifier. The `AuthState` was extended such that it includes an `AuthStatus` which can be `UNAUTHENTICATED`, `AUTHENTICATING`, `AUTHENTICATED`, `EXPIRED`. The lifecycle of the auth status is depicted below:

```mermaid
stateDiagram-v2
    direction LR
    [*] --> Unauthenticated
    Unauthenticated --> Authenticating
    Authenticating --> Unauthenticated
    Authenticating --> Authenticated
    Authenticated --> Expired
    Expired --> [*]
```
Note: in general, the JWT should be renewed before it expires. Renewing the JWT is currently not supported and will be the subject of a follow-up PR.

**_Questions_**

- Electric supports top-level JWT claims and namespaced claims. **Should `exp` always be a top-level claim?** (currently implemented that way)
- **Should the `exp` claim always be present?** It seems so because it is listed in `required_claims` but it seems that Electric does not complain when that claim is omitted. Either, it should be required and Electric must complain if it is not present, or, it is not required and Electric must use a default expiration.

@alco can probably answer these questions

--------

EDIT: Added support for renewing the JWT. This fixes VAX-1432 and VAX-1435.

alco commented:
@kevin-dp Hey, nice work on the Elixir code!

I've simplified it a bit and pushed a bunch of commits to the branch. See commit messages for explanation.

Regarding your question about the `exp` claim, it is a standard claim, so it is always at the top level. I'm not sure how you got the server to accept a token without this claim in the `secure` mode. There is a unit test that verifies server's handling of this claim - https://github.com/electric-sql/electric/blob/e11501d8d3eb7c77ccf49451893926e8adc09ae8/components/electric/test/electric/satellite/auth/secure_test.exs#L179-L201.

balegas reviewed (commented):
What is the behaviour when the token expires? Given that expired is a terminal state and that electric closes the connection, the only option for the client is to close too.

Right now, we leave the client in a limbo, expecting that the developer subscribes to the auth state changes, closes the client, fetches a new token and starts again.

I think for the scope of this patch we should just close the client and throw an error. We should improve any examples that use auth to handle the token expiration gracefully. Note that is a serious concern, given that the website breaks hard when we don't handle errors -- talk with me if you find there is nothing to change regarding this.

I imagine the code evolving once we know how we refresh the token. I have no objections if the team feels that we should learn that before committing this code.

kevin-dp commented:
@balegas I modified the implementation as discussed over our call some time ago. User's now get notified of the JWT expiration through the connectivity status which is set to `disconnected` on expiration together with an error message, i.e. the connectivity state now has the following type:

```ts
export type ConnectivityStatus = 'available' | 'connected' | 'disconnected'
export type ConnectivityState = {
  status: ConnectivityStatus
  error?: string // reason for `disconnected` status
}
```

> I think for the scope of this patch we should just close the client and throw an error.

I modified the implementation such that on JWT expiration, we close the Satellite client and process.
I'm not sure about closing the process. Perhaps it would be enough to close the client if the user has a way to re-connect with a new token (without having to re-electrify the database). Re-connecting with a new token should work as long as it has the same user ID. If that's possible, we can envision re-connection and renewal to look like this:
```ts
connect(token: AuthToken) // only possible if client is disconnected, must preserve user ID
renew(token: AuthToken) // only possible if client is connected, must preserve user ID
```

kevin-dp commented:
I added the possibility to renew the JWT. The new state diagram looks like this:

```mermaid
stateDiagram-v2
    direction LR
    [*] --> Unauthenticated
    Unauthenticated --> Authenticating
    Authenticating --> Unauthenticated
    Authenticating --> Authenticated
    Authenticated --> Authenticated: renew
    Authenticated --> Expired
    Expired --> [*]
```
The JWT should be renewed before it expires. If the JWT expires, the Satellite client and process are closed; the user will need to re-electrify the database with a new JWT.

alco reviewed (approved):
Good job!

I've pushed a few tweaks to the Elixir code.

kevin-dp commented on components/electric/lib/electric/satellite/protocol.ex:
```diff
@@ -312,6 +312,44 @@ defmodule Electric.Satellite.Protocol do
+    else
+      {:error, %Electric.Satellite.Auth.TokenError{message: message}} ->
+        Logger.warning("Client authentication failed: #{message}")
+        {:error, %SatErrorResp{error_type: :AUTH_REQUIRED}}
```
Totally! Will fix this.

kevin-dp commented:
Modified the implementation as suggested by @balegas such that the user does not need to re-electrify when the token expires but can reconnect with a new token. The new state diagram looks like this:

```mermaid
stateDiagram-v2
    direction LR
    [*] --> Unauthenticated
    Unauthenticated --> Authenticating
    Authenticating --> Unauthenticated
    Authenticating --> Authenticated
    Authenticated --> Authenticated: renew
    Authenticated --> Expired
    Expired --> Authenticated: reconnect
```

Example usage:

```ts
function fetchJWT() {
  // oracle that fetches your JWT somehow
}

// Electrify the DB
const electric = await electrify(conn, schema)
// Fetch our JWT
// this example assumes it expires in > 1 hour
const token = fetchJWT()
// Connect to Electric
await electric.connect(token)
// Renew the JWT every hour
const oneHour = 60 * 60 * 1000
let stopRenewing = renewPeriodically(electric, oneHour)

// Subscribe to connectivity changes to detect JWT expiration
electric.notifier.subscribeToConnectivityStateChanges(async (x) => {
  if (x.connectivityState.status === 'disconnected' && x.connectivityState.error === 'JWT expired') {
    console.log("JWT expired, reconnecting...")
    stopRenewing() // NOTE: the connectivity state change event is async and is fired after the socket to Electric is closed. Between the socket closing and this event firing, we may have tried to renew the token which will fail
    const newToken = fetchJWT()
    await electric.connect(newToken)
    console.log("connection restored")
    stopRenewing = renewPeriodically(electric, oneHour)
  }
})

// Renews the JWT periodically
// and returns a function that can be called to stop renewing
function renewPeriodically(electric: Electric, ms: number) {
  const id = setInterval(async () => {
    // Renew the JWT
    const renewedToken = fetchJWT()
    await electric.renew(renewedToken)
    console.log("Renewed JWT")
  }, ms)
  return () => {
    clearInterval(id)
  }
}
```

balegas commented on docs/api/clients/typescript.md:
How shall we handle this situation? 
- we don't want blocking merging to main
- we want to write docs along with PRs

I believe we're not at the point where we could simply make another patch release. In this particular instance, probably there is no harm...

Shall we maintain a open PR with doc changes for the next release? On every PR that affects docs, we do a commit with doc changes and before merge we remove it from PR and cherry pick into the docs PR?
It's a bit of work, but is easy for review and we could always point to latest docs

kevin-dp commented on clients/typescript/src/frameworks/react/hooks/useConnectivityState.ts:
```diff
@@ -17,11 +17,11 @@ const STATES: {
+  disconnected: { status: 'disconnected' },
 }
-const VALID_STATES = Object.values(STATES)
+const VALID_STATUSES = Object.values(STATES).map((s) => s.status)
```
I extended the connectivity state with an optional `reason` field such that we can signal the reason why the client is disconnected. For example, because the JWT expired. I now this can no longer happen since we said JWT should only be used for authentication, but it will happen later when we introduce sessions and want to signal that the client is disconnected because the session expired.

See here the relevant types:
```ts
export type ConnectivityStatus = 'available' | 'connected' | 'disconnected'
export type ConnectivityState = {
  status: ConnectivityStatus
  reason?: SatelliteError // reason for `disconnected` status
}
```

So i modified the `STATES` according to the new `ConnectivityState` type.
The disconnected state does not use the `reason` field because we only know that at runtime on an actual disconnection.

kevin-dp replied:
Wrt to your question. No, `STATES.available`, `STATES.connected`, and `STATES.disconnected` are just convenience variables that represent those respective states such that we can e.g. return `STATES.connected` where needed instead of having to return `{ status: 'connected' }` everywhere we want to return the connected state.

alco replied:
Thanks! That clears it up for me.

samwillis reviewed (approved):
LGTM 👍
